src/librustc_typeck/check/mod.rs

```diff
@@ -971,8 +986,9 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
                             fn_sig: ty::FnSig<'tcx>,
                             decl: &'gcx hir::FnDecl,
                             fn_id: ast::NodeId,
-                            body: &'gcx hir::Body)
-                            -> FnCtxt<'a, 'gcx, 'tcx>
+                            body: &'gcx hir::Body,
+                            can_be_generator: bool)
```

Comment:
`can_be_generator` is false for functions. So we don't treat that as a generator even if the body has yields statements.